aana/deployments/task_queue_deployment.py

```diff
@@ -61,81 +58,82 @@ async def apply_config(self, config: dict[str, Any]):
         """
-
-        async def handle_task(task_id: str):
-            """Process a task."""
-            # Fetch the task details
-            task = self.task_repo.read(task_id)
-            # Initially set the task status to RUNNING
-            self.task_repo.update_status(task_id, TaskStatus.RUNNING, 0)
-            try:
-                # Call the endpoint asynchronously
-                out = await self.handle.call_endpoint.remote(task.endpoint, **task.data)
-                # Update the task status to COMPLETED
-                self.task_repo.update_status(task_id, TaskStatus.COMPLETED, 100, out)
-            except Exception as e:
-                # Handle the exception and update the task status to FAILED
-                error_response = custom_exception_handler(None, e)
-                error = orjson.loads(error_response.body)
-                self.task_repo.update_status(task_id, TaskStatus.FAILED, 0, error)
-
-        def run_handle_task(task_id):
-            """Wrapper to run the handle_task function."""
-            run_async(handle_task(task_id))
-
-        def is_thread_pool_full():
-            """Check if the thread pool has too many tasks.
-
-            We use it to stop assigning tasks to the thread pool if it's full
-            to prevent the thread pool from being overwhelmed.
-            We don't want to schedule all tasks from the task queue (could be millions).
-            """
-            return (
-                self.thread_pool._work_queue.qsize()
-                > aana_settings.task_queue.num_workers * 2
-            )
+        running_task_ids: list[str] = []
+        handle = None
+
+        active_tasks = self.task_repo.get_active_tasks()
+        for task in active_tasks:
+            if task.status == TaskStatus.RUNNING:
+                running_task_ids.append(task.id)
+            if task.status == TaskStatus.ASSIGNED:
+                self.task_repo.update_status(
+                    task_id=task.id,
+                    status=TaskStatus.NOT_FINISHED,
+                    progress=0,
+                )
 
         while True:
             if not self._configured:
                 # Wait for the deployment to be configured.
                 await asyncio.sleep(1)
                 continue
 
-            # Remove completed tasks from the futures dictionary
-            for task_id in list(self.futures.keys()):
-                if self.futures[task_id].done():
-                    del self.futures[task_id]
-
-            if is_thread_pool_full():
-                # wait a bit to give the thread pool time to process tasks
+            # Remove completed tasks from the list of running tasks
+            running_task_ids = self.task_repo.remove_completed_tasks(running_task_ids)
+
+            # Check for expired tasks
+            execution_timeout = aana_settings.task_queue.execution_timeout
+            expired_tasks = self.task_repo.get_expired_tasks(execution_timeout)
```

Comment:
We need to have a mechanism to cancel the task, so the loop does not flood with some unfinished futures (In case the specify endpoint have a problem and does not finished for any reason)

Reply:
Forexample if we can implement a timeout in the execute_task function too.

Reply:
Added deployment response cancellation.